## Welcome to Team "Epigenomics" GitHub Page <br>
Our team's interests center on the analysis of epigenomic and genomic datasets.  Thus, we are interested in analyzing a variety of large datasets, often derived from next generation sequencing.  These datasets include RNA-seq, ChIP-seq, ATAC-seq, RRBS data, and more.  [Managing this variety of datasets and formats](https://github.com/kfreij95/Group-A-Capstone-I/blob/main/Data-Files/datafilesreadme.md), as well as the analysis pipelines needed is something we hope to use the tools provided by Cyverse for. 

## Meet the Team <br>

| |  |
|---|---|
| <img src="https://www.uab.edu/cas/biology/images/images/people/faculty/nicole-riddle.jpg"> | [**Nicole C. Riddle, PhD**](https://github.com/riddlenc)<br/> Nicole is an Associate Professor in the Department of Biology at the University of Alabama at Birmingham. |
| <img src="https://www.uab.edu/cas/biology/images/images/people/faculty/peggy-biga.jpg"> | [**Peggy R. Biga, PhD**](https://github.com/peggybiga)<br/> Peggy is an Associate Professor in the Department of Biology at the University of Alabama at Birmingham. |
| <img src="https://www.uab.edu/cas/biology/images/images/people/grad_students/khalid-freij.jpg"> | [**Khalid W. Freij**](https://github.com/kfreij95)**,** [**PhD Student**](https://www.researchgate.net/profile/Khalid-Freij)<br/> Khalid is a PhD Student in the Department of Biology at the University of Alabama at Birmingham. |
| <img src="https://imgur.com/refmO39.jpg"> | [**Serhat Turkmen, PhD**](https://github.com/s3rhat)<br/> Serhat is a Postdoctoral Fellow in the Department of Biology at the University of Alabama at Birmingham. |
| <img src="https://i2.wp.com/publichealth.berkeley.edu/wp-content/uploads/2020/08/phil-c-98.jpg?w=900&ssl=1" width="300"> | [**Philip Collender**](https://github.com/pcollender)**,** [**MPH**](https://scholar.google.com/citations?user=dp7nggQAAAAJ&hl=en)<br/> Philip is an academic researcher at the University of California, Berkeley School of Public Health, Environmental Health Sciences division. |


## Lessons learned from FOSS <br>
Reflections from our group members on on tools or concepts learned through FOSS and how we might apply them to their field of study. Specifically, we address the following questions:<br>
    - Which tool could you apply right away and why?<br>
    - What is your greatest barrier to these tools?<br>
    - What tools would you wish we covered more in depth?<br>

### Peggy <br>
To begin with, I have learned to outline my data management strategies and use an open platform, like CyVerse, in that strategy. The data management plan tool is super helpful when building a research data plan in terms of data storage, data sharing, and analysis workflows. I am currently applying these concepts in two grant proposals I am writing, and will engage the students in my research lab to also incorporate the CARE and FAIR dta standards to their work. I plan to utilize CyVerse to work collaboratively on several projects that are on going, where we will share data, share data analysis pipelines that will include workflows for RNAseq and RRBS analyses. This will allow all collaborators on the projects to use a standardized workflow, with version control. This will require me spending more time learning about and understanding package managers and containers more, and the barrier is finding the time and bandwidth to devote to this!

### Phil <br>
The FOSS course broadened my awareness of tools and philosophy of scientific reproducibility and data stewardship. I had some baseline familiarity with Github, command line applications, and cloud computing, but it was very useful for me to see the data management cycle laid out explicitly, and more advanced tools such as Docker and Github pages were new to me. I think I still need some time and effort to understand best practices for applying containers, and even for optimally managing Github repos, but I've already began to apply some of the concepts in small collaborative projects, including developing a pipeline for decentralized linking of firearm-related injury records to geographic locations ([link](https://github.com/pcollender/Firearm-injury-geocoding)). For my personal development with these tools, I think I could use more intensive training on methods for managing Github repositories effectively, as well as some more applied work using Docker. In particular, I'm interested in learning how to make Docker containers that find an optimal balance between reproducibility and flexibility, as I would like to use them as a tool to facilitate deployment of customized software in a collaborative setting.


### Khalid <br>
I have been able to apply a data management plan regarding how I will be storing my RRBS and RNASeq data I'm set to receive soon. In incorporating these ideas like CARE and FAIR data standards, I've become more conscious with how to not only handle my data but to create a standard naming structure for datasets to be better organized for all parties involved now and other scientists later. The programs and command lines don't worry me as much as when I started a year ago, but with the ability to transfer my knowledge to different software, the only barrier(s) would be learning the software in question and how it functions. Another is getting lost (in a good way!) in the coding and losing track of time. Using this class, the "buffet" example was pretty accurate. I've learned a lot, and while it is more on how to get started on several programs, that's all you need to get started on your own or even take more advanced courses like the container courses.

### Nicole <br>
The discussions about data management, data integrity, and FAIR data standards have been very useful. I have been able to incorporate these ideas into a data management plan for a grant to be submitted this spring.  In addition, I have made more conscious decisions on how to handle data, file naming, etc for a new project that we just started in the lab.  Applying these concepts will help me encourage my laboratory to strive to meet FAIR data standards and apply Open Science principles whenever possible.  In terms of computational tools, I would like to use Cyverse to develop and implement shared data analysis pipelines for genomics data analysis to have one standard for all lab personnel and to prevent having multiple parallel pipelines implemented by different people.  To achieve that, I need to big deeper into the tools provided by Cyverse to do that, learn about pipeline managers and Docker.  For me, the main barrier is lack of time.  

### Serhat <br>

